### PR TITLE
Normalized xpointers, complete incomplete tables

### PR DESCRIPTION
**Pull request by poire-z**
See individual commit messages for details.
Some additional notes:

`(Upstream) createXPointer/toString: normalized xpointers`
Bits from 3 commits by @pkb from https://github.com/buggins/coolreader/pull/125, stuff I never dared doing :) but glad we finally have that: XPointers without the boxing elements, that should survive changes in boxing/CSS `display:` property values.
Note that I didn't pick their own CRHistFile xpointers migration and gDOMVersionRequested update (that I initially discussed there with them), as we are actually using that CRHistfile in [this little bit of frontend code](https://github.com/koreader/koreader/blob/071c22f94eb9be2c8a900cd38bdfc07bdc3e58ea/frontend/apps/reader/modules/readerrolling.lua#L831-L852) that deal with orientation/window size change - and given the hard time I had to make that work, and the fact that CRHistFile wasn't really easy to understand, best to not take risks and leave it untouched.

`toStringUsingNames(): rename to toStringV2()`
`toStringV2(): output [1] when first but not single`
`DOM_VERSION_WITH_NORMALIZED_XPOINTERS 20200223`
are small adaptations to that Upsteam code.

`CSS/List items: skip boxing elements when walking` is, I hope, the last needed bit to finally get transparent boxing elements: instead of iterating all the parent's children in a possibly-boxed-DOM, and so possibly missing some of them, we start from the first child and use `getUnboxedNextSibling()` to iterate all original siblings by walking up and down boxing elements.
This is a bit more expensive, but not really noticable on most books (but on the nwt-E crappy EPUB from https://github.com/koreader/crengine/issues/277#issuecomment-513918425, there is a 25% increase of load & rerendering times as it abuses `:first-child/nth-child()` pseudoclasses - and it's not because of the next commit `CSS: parse pseudoclass nth_child(3n+2)`, same 25% without it).

`Tables: complete incomplete tables` was discussed and detailed alongside https://github.com/koreader/crengine/pull/325#issuecomment-580438763. This may reveal new content that was previously just discarded/invisible.
Follow rules from https://www.w3.org/TR/CSS22/tables.html#anonymous-boxes and https://www.w3.org/TR/css-tables-3/#fixup except that we do _3) Generate missing parents_ before _2) Generate missing child wrappers_.

This comes with:
`DOM_VERSION_CURRENT 20200223`
`DOM_VERSION_WITH_NORMALIZED_XPOINTERS 20200223`
so only newly opened books will benefit from them.
But I'll bump this with some frontend code to propose a migration/upgrade, or do it without asking when no risk. Still need a bit of thinking on this.

image

image

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/crengine/328)

### CODE REVIEW

File: crengine/src/lvtinydom.cpp [hunk 1]
```diff
@@ -461,9 +461,11 @@ struct SimpleCacheFileHeader
 {
     char _magic[CACHE_FILE_MAGIC_SIZE] = { 0 }; // magic
     lUInt32 _dirty;
+    lUInt32 _dom_version;
     SimpleCacheFileHeader( lUInt32 dirtyFlag ) {
         memcpy( _magic, _compressCachedData ? COMPRESSED_CACHE_FILE_MAGIC : UNCOMPRESSED_CACHE_FILE_MAGIC, CACHE_FILE_MAGIC_SIZE );
         _dirty = dirtyFlag;
+        _dom_version = gDOMVersionRequested;
```
Comment (Frenzie) [inline crengine/src/lvtinydom.cpp]: Good idea. :+1:

**Comment by poire-z** [comment]
Rebasing to just re-order 2 commits (that I had badly ordered when grouping them by topic):

`toStringV2(): output [1] when first but not single` depends on stuff added by `CSS/List items: skip boxing elements when walking`

**Comment by poire-z** [comment]
There seems to be a case where xpointers include some bits of our boxing elements, may be when that boxing element is the last.

When bookmarking a page that starts with some inline box containing an image:
```
["page"] = "/body/DocFragment[16]/body/div[2]/div[3]/div[1]/div[2]/inlineBox.0",
["text"] = "Page 1 in Couverture @ 2020-02-25 22:03:32",
```
which resolves badly to page 1 :(
When selecting and highlighting some text in that inline box, we get:
```
["page"] = "/body/DocFragment[16]/body/div[2]/div[3]/div[1]/div[2]/div[1]/span/text().14",
["highlighted"] = true,
["pos1"] = "/body/DocFragment[16]/body/div[2]/div[3]/div[1]/div[2]/div[1]/span/text().33",
["pos0"] = "/body/DocFragment[16]/body/div[2]/div[3]/div[1]/div[2]/div[1]/span/text().14"
```

**Comment by poire-z** [comment]
I think this should solve that ^:
```diff
@@ -8458,10 +8500,18 @@ lString16 ldomXPointer::toStringV2()
         return path;
     ldomNode * node = getNode();
     int offset = getOffset();
-    if ( offset >= 0 ) {
-        path << "." << fmt::decimal(offset);
-    }
     ldomNode * p = node;
+    if ( !node->isBoxingNode() ) {
+        if ( offset >= 0 ) {
+            path << "." << fmt::decimal(offset);
+        }
+    }
+    else {
+        if ( offset < p->getChildCount() )
+            p = p->getChildNode(offset);
+        else
+            p = p->getParentNode();
+    }
     ldomNode * mainNode = node->getDocument()->getRootNode();
     while (p && p!=mainNode) {
         ldomNode * parent = p->getParentNode();
```
We'd then get:
```
["page"] = "/body/DocFragment[16]/body/div[2]/div[3]/div[1]/div[2]/div[1]"
instead of
["page"] = "/body/DocFragment[16]/body/div[2]/div[3]/div[1]/div[2]/inlineBox.0",
```

**Comment by Frenzie** [comment]
Alright, nice. :-)

**Comment by poire-z** [comment]
Cross-referencing: `CSS/List items: skip boxing elements when walking` does what I mentionned in https://github.com/koreader/koreader/pull/3934#issuecomment-387098774

**Pull request merged**